src/rgw/rgw_bucket_sync_cache.h

```diff
@@ -0,0 +1,147 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab ft=cpp
+
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2020 Red Hat, Inc
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ */
+
+#pragma once
+
+#include <boost/intrusive/set.hpp>
+#include <boost/intrusive/list.hpp>
+#include "rgw_data_sync.h"
+
+namespace rgw::bucket_sync {
+
+// per bucket-shard state cached by DataSyncShardCR
+struct State {
+  // the source bucket shard to sync
+  rgw_bucket_shard key;
+  // current sync obligation being processed by DataSyncSingleEntry
+  std::optional<rgw_data_sync_obligation> obligation;
+  // highest timestamp applied by all sources
+  ceph::real_time progress_timestamp;
+};
+
+class Cache;
+
+namespace bi = boost::intrusive;
+
+// a recyclable cache entry
+struct Entry {
+  State state;
+  bi::set_member_hook<> key_hook;
+  bi::list_member_hook<> lru_hook;
+  uint32_t pins = 1; // not eligible for recycling while pins > 0
+};
+
+// a State handle that unpins itself from the Cache on destruction
+struct Handle {
+  std::shared_ptr<Cache> cache;
+  Entry* entry = nullptr;
+ public:
+  Handle() noexcept;
+  Handle(std::shared_ptr<Cache> cache, Entry* entry) noexcept;
+  Handle(Handle&& o) noexcept;
+  Handle& operator=(Handle&& o) noexcept;
+  Handle(const Handle&) = delete;
+  Handle& operator=(const Handle&) = delete;
+  ~Handle();
+
+  explicit operator bool() const noexcept { return entry; }
+  State& operator*() const noexcept { return entry->state; }
+  State* operator->() const noexcept { return &entry->state; }
+};
+
+// a state cache for entries within a single datalog shard
+class Cache : public std::enable_shared_from_this<Cache> {
```

Comment:
There's a lot of stuff, but some is not great and some is even strange.  I don't want to inherit !great or strange, at any rate.

Reply:
thanks, intrusive_lru.h looks extremely close to what i did here. no locks/atomics, intrusive hooks, etc.

its handles don't keep the cache itself referenced (which we need here, because coroutine destruction order isn't deterministic), but i should be able to add that with a wrapper. i'll give it a try

Reply:
updated with intrusive_lru, and it did simplify quite a bit